File: take_u_forward/graph/28_Word Ladder II.cpp
class Solution {
  public:
    vector<vector<string>> findSequences(string beginWord, string endWord,
                                         vector<string>& wordList) {
        
        unordered_set<string> st(wordList.begin(), wordList.end());

        queue<vector<string>> q;
        q.push({beginWord});

        vector<string> usedOnLevel;
        usedOnLevel.push_back(beginWord);

        int level = 1;
        vector<vector<string>> ans;

        while (!q.empty()) {
            vector<string> seq = q.front();
            q.pop();

            // If we moved to next level
            if (seq.size() > level) {
                level = seq.size();
                for (string &word : usedOnLevel) {
                    st.erase(word);
                }
                usedOnLevel.clear();
            }

            string word = seq.back();

            // If target reached
            if (word == endWord) {
                if (ans.empty()) {
                    ans.push_back(seq);
                } else if (ans[0].size() == seq.size()) {
                    ans.push_back(seq);
                }
                continue; // do not expand further
            }

            for (int i = 0; i < word.size(); i++) {
                char original = word[i];

                for (char c = 'a'; c <= 'z'; c++) {
                    word[i] = c;
                    if (st.count(word)) {
                        seq.push_back(word);
                        q.push(seq);
                        usedOnLevel.push_back(word);
                        seq.pop_back();
                    }
                }

                word[i] = original;
            }
        }

        return ans;
    }
};

File: take_u_forward/graph/28_Word Ladder II.java
import java.util.*;

class Solution {
    public ArrayList<ArrayList<String>> findSequences(String startWord,
                                                      String targetWord,
                                                      String[] wordList) {

        // Result list
        ArrayList<ArrayList<String>> ans = new ArrayList<>();

        // Store words in a HashSet for fast lookup
        HashSet<String> set = new HashSet<>(Arrays.asList(wordList));

        // Queue to store sequences
        Queue<ArrayList<String>> q = new LinkedList<>();
        q.offer(new ArrayList<>(List.of(startWord)));

        // Words used at current BFS level
        ArrayList<String> usedOnLevel = new ArrayList<>();
        usedOnLevel.add(startWord);

        int level = 1;

        while (!q.isEmpty()) {
            ArrayList<String> seq = q.poll();

            // Level change → remove previously used words
            if (seq.size() > level) {
                level = seq.size();
                for (String word : usedOnLevel) {
                    set.remove(word);
                }
                usedOnLevel.clear();
            }

            String word = seq.get(seq.size() - 1);

            // If target reached
            if (word.equals(targetWord)) {
                if (ans.isEmpty()) {
                    ans.add(new ArrayList<>(seq));
                } else if (ans.get(0).size() == seq.size()) {
                    ans.add(new ArrayList<>(seq));
                }
                continue;
            }

            // Try all transformations
            char[] wordArr = word.toCharArray();
            for (int i = 0; i < wordArr.length; i++) {
                char original = wordArr[i];

                for (char c = 'a'; c <= 'z'; c++) {
                    wordArr[i] = c;
                    String newWord = new String(wordArr);

                    if (set.contains(newWord)) {
                        seq.add(newWord);
                        q.offer(new ArrayList<>(seq));
                        usedOnLevel.add(newWord);
                        seq.remove(seq.size() - 1);
                    }
                }
                wordArr[i] = original;
            }
        }

        return ans;
    }
}